sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+
+            let mut pkt = Vec::<u8>::new();
+            pkt.write_u8(Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO)?;
+            pkt.write_u8(Self::GPIO_MONITORING_READ)?;
+            pkt.write_u8(pin_names.len().try_into()?)?;
+            for pin_name in &pin_names {
+                pkt.write_u8(pin_name.len().try_into()?)?;
+                pkt.extend_from_slice(pin_name.as_bytes());
+            }
+            self.inner
+                .usb_device
+                .borrow()
+                .write_bulk(cmsis_interface.out_endpoint, &pkt)?;
+
+            let mut resp = RspGpioMonitoringHeader::new();
+            let mut bytecount = 0;
+            while bytecount < GPIO_MONITORIN_HEADER_SIZE {
+                let read_count = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut resp.as_bytes_mut()[bytecount..][..64],
+                )?;
+                ensure!(
+                    read_count > 0,
+                    TransportError::CommunicationError("Truncated GPIO response".to_string())
+                );
+                bytecount += read_count;
+            }
+            ensure!(
+                resp.encapsulation_header == Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO,
+                TransportError::CommunicationError(
+                    "Unrecognized CMSIS-DAP response to GPIO request".to_string()
+                )
+            );
+            match resp.status {
+                0 => (),
+                n => bail!(TransportError::CommunicationError(format!(
+                    "HyperDebug error: {}",
+                    n
+                ))),
+            }
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
+
+            let mut databytes: Vec<u8> = Vec::new();
+            databytes.extend_from_slice(&resp.data[..bytecount - GPIO_MONITORIN_HEADER_SIZE]);
+
+            while databytes.len() < skip_bytes + resp.transcript_size as usize {
+                let original_length = databytes.len();
+                databytes.resize(original_length + 64, 0u8);
+                let c = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut databytes[original_length..],
+                )?;
+                databytes.resize(original_length + c, 0u8);
+            }
+
+            // Figure out how many of the low bits are used for storing the index of the signal
+            // hanving changed. (If only one signal, no bits are used, if two signals, then one
+            // bit is used, if three or four, then two bits are used, etc.)
+            let mut signal_bits = 0;
+            while (pin_names.len() - 1) >> signal_bits != 0 {
+                signal_bits += 1;
+            }
+            let signal_mask = (1u64 << signal_bits) - 1;
+
+            let mut cur_time: u64 = resp.start_timestamp;
+            let mut cur_level = Vec::new();
```

Comment:
Yes, I chose `Vec<bool>` because I thought that the code in the loop would become cleaner, but it seems that it is causing more trouble than it is saving.  I have changed the code to use a bitmask in the format of `start_levels`.